third_party/libusb/naclport/src/chrome_usb/chrome-login-state-hook.js

```diff
@@ -0,0 +1,171 @@
+        'chrome.loginState API is not available. This app might require a '
+        + 'newer version of Chrome.');
+    this.hookIsReadyResolver_.reject();
+  }
+};
+
+/** @const */
+const ChromeLoginStateHook = GSC.Libusb.ChromeLoginStateHook;
+
+goog.inherits(ChromeLoginStateHook, goog.Disposable);
+
+/**
+ * @type {!goog.log.Logger}
+ * @const
+ */
+ChromeLoginStateHook.prototype.logger = GSC.Logging.getScopedLogger(
+  'Libusb.LoginStateHook');
+
+/** @override */
+ChromeLoginStateHook.prototype.disposeInternal = function() {
+  if (chrome.loginState) {
+    chrome.loginState.onSessionStateChanged.removeListener(
+        this.boundOnGotSessionState_);
+  }
+
+  this.logger.fine('Disposed');
+
+  ChromeLoginStateHook.base(this, 'disposeInternal');
+};
+
+/**
+ * @return {function(string, !Array): !Array}
+ * @public
+ */ 
+ChromeLoginStateHook.prototype.getRequestSuccessHook = function() {
+  return this.requestSuccessHook_.bind(this);
+};
+
+/**
+ * @return {!goog.Promise}
+ * @public
+ */ 
+ChromeLoginStateHook.prototype.getHookReadyPromise = function() {
+  return this.hookIsReadyResolver_.promise;
+};
+
+ /**
+ * @param {!chrome.loginState.ProfileType} profileType
+ * @private
+ */
+ChromeLoginStateHook.prototype.onGotProfileType_ = function(profileType) {
+  goog.asserts.assert(chrome.loginState);
+  if (profileType === chrome.loginState.ProfileType.USER_PROFILE) {
+    chrome.loginState.getSessionState(this.boundOnGotSessionState_);
+    chrome.loginState.onSessionStateChanged.addListener(
+        this.onSessionStateChanged_.bind(this));
+  } else {
+    this.hookIsReadyResolver_.resolve();
+  }
+};
+
+/**
+ * @param {!chrome.loginState.SessionState} sessionState
+ * @private
+ */
+ChromeLoginStateHook.prototype.onGotSessionState_ = function(
+    sessionState) {
+  goog.asserts.assert(chrome.loginState);
+  if (sessionState === chrome.loginState.SessionState.IN_LOCK_SCREEN
+      && !this.simulateDevicesAbsent_) {
+    this.logger.info("No longer showing USB devices.");
+    this.simulateDevicesAbsent_ = true;
+  } else if (sessionState !== chrome.loginState.SessionState.IN_LOCK_SCREEN
+             && this.simulateDevicesAbsent_) {
+    this.logger.info("Showing USB devices.");
+    this.simulateDevicesAbsent_ = false;
+  }
+  this.hookIsReadyResolver_.resolve();
+};
+
+/**
+ * @private
+ */
+ChromeLoginStateHook.prototype.onSessionStateChanged_ = function() {
+  chrome.loginState.getSessionState(this.boundOnGotSessionState_);
+};
+
+/**
+ * @param {string} functionName chrome.usb function that was called
+ * @param {!Array} callResults
+ * @return {!Array} new callResults
+ * @private
+ */
+ChromeLoginStateHook.prototype.requestSuccessHook_ = function(functionName, 
+                                                              callResults) {
+  if (this.simulateDevicesAbsent_ &&
+      (functionName === 'getDevices' ||
+       functionName === 'getConfigurations')) {
+    callResults[0] = [];
```

Comment:
This still modifies the value of the argument, which in general is quite error-prone and I think should only be used for a clear reason and with a specific documentation.

I don't remember - do we have more than one item in |callResults| for these calls? If not, we could just do "return []", otherwise goog.array.clone() or something similar could help.